### PR TITLE
[SPARK-21422][BUILD] Depend on Apache ORC 1.4.0

### PR DESCRIPTION
**Pull request by dongjoon-hyun**
## What changes were proposed in this pull request?

Like Parquet, this PR aims to depend on the latest Apache ORC 1.4 for Apache Spark 2.3. There are key benefits for Apache ORC 1.4.

- Stability: Apache ORC 1.4.0 has many fixes and we can depend on ORC community more.
- Maintainability: Reduce the Hive dependency and can remove old legacy code later.

Later, we can get the following two key benefits by adding new ORCFileFormat in SPARK-20728 (#17980), too.
- Usability: User can use ORC data sources without hive module, i.e, -Phive.
- Speed: Use both Spark ColumnarBatch and ORC RowBatch together. This will be faster than the current implementation in Spark.

## How was this patch tested?

Pass the jenkins.

### CODE REVIEW

**Comment by dongjoon-hyun** [comment]
This aims to reduce the review scope for #17980 .
cc @kiszk .

**Comment by dongjoon-hyun** [comment]
Hi, @rxin , @srowen , @sameeragarwal , @cloud-fan , @hvanhovell , @gatorsmile , @ueshin , @viirya , @kiszk .

Could you review this small PR about depedency change?

This is a start of upgrade to Apache ORC in order to reduce the old Hive dependency in Apache Spark 2.3 for the following issues.

- SPARK-20901 Feature parity for ORC with Parquet
- SPARK-20682 Support a new faster ORC data source based on Apache ORC
- SPARK-20728 Make ORCFileFormat configurable between sql/hive and sql/core
- SPARK-16060 Vectorized Orc Reader

I've heard that Apache Spark will not drop ORC data source from @sameeragarwal . If then, could we move forward a small step like this?

**Comment by dongjoon-hyun** [comment]
Retest this please.

**Comment by dongjoon-hyun** [comment]
Retest this please.

**Comment by dongjoon-hyun** [comment]
Retest this please

**Comment by dongjoon-hyun** [comment]
Hi, @liancheng , @zhzhan , @rxin , @marmbrus .
I'm pining you since you worked on #6194 before.

**Comment by rxin** [comment]
Why are we adding this to core? Why not just the hive module?

**Comment by kiszk** [comment]
Can we add any smaller code to use this, too?

**Comment by dongjoon-hyun** [comment]
Thank you for review, @rxin .
We can use ORC like Parquet now. Parquet is inside `sql/core`, not `sql/hive`.

**Comment by dongjoon-hyun** [comment]
Thank you for review, @kiszk .
The example may be #17980 , #17924, and #17943 .
If possible, in this PR, I want to focus on only `Dependency on ORC` issue.

**Comment by mridulm** [comment]
LGTM, great to see progress on ORC support.

**Comment by rxin** [comment]
To the best of my knowledge almost everybody runs with Hive anyway and the vast majority of users that run ORC are Hive users. In hindsight we probably should have put most of the data source dependencies as separate packages similar to Presto.

**Comment by dongjoon-hyun** [comment]
I agree with the following, but this does not block those users. This is only better than putting the dependency on Hive because it also supports more the other users who are using ML and storage, too. In addition, when we refactor the data source dependecies, this will help the refactoring as clean as Parquet.
> To the best of my knowledge almost everybody runs with Hive anyway and the vast majority of users that run ORC are Hive users.

**Comment by rxin** [comment]
Why don't we then create a separate orc module? Just copy a few of the files over?

**Comment by dongjoon-hyun** [comment]
Until now, I think ORC is the same with most of other data sources(CSV, JDBC, JSON, PARQUET, TEXT) which live inside `sql/core` now. If that is an architectural plan of Apache Spark 2.3, I will. Are we going to move out all data sources into separate modules, e.g., `datasources/parquet`, in timeframe of Spark 2.3?

Or, is there any other reason I don't catch here?

**Comment by rxin** [comment]
I just checked the dependency size. They look pretty reasonable, roughly 2 MBs in total (although I do worry in the future whether ORC would bring in a lot more jars).

cc @omalley any guidance on this topic?

**Comment by dongjoon-hyun** [comment]
Hi, @rxin 

Since ORC 1.4.0, ORC community provides small shaded jar files to improve usability in general purposes. This PR uses the followings.
- orc-core-1.4.0-nohive.jar (1.4MB)
- orc-mapreduce-1.4.0-nohive.jar (739KB)

The size is due to including the [followings](https://github.com/apache/orc/blob/master/java/pom.xml#L258-L259).
- com.google.protobuf:protobuf-java
- org.apache.hive:hive-storage-api

In terms of the number of files,
- ORC (354 files)
- ProtoBuf (247 files)
- Hive Storage API (92 files)

The bottom line is there are still some source codes come from `org.apache.hive` namespace originally. So, I'm wondering if this is the reason why you want to put this into `sql/hive` module still and want to copy source codes instead of using this shaded jar?

**Comment by omalley** [comment]
@rxin The ORC core library's dependency tree is aggressively kept as small as possible. I've gone through and excluded unnecessary jars from our dependencies. I also kick back pull requests that add unnecessary new dependencies.

**Comment by dongjoon-hyun** [comment]
Thank you, @omalley .

@rxin . I think we had better depend on Apache ORC libraries as is in this PR.

**Comment by dongjoon-hyun** [comment]
@rxin . How can I proceed this PR now? Could you give me some advice again?

**Comment by dongjoon-hyun** [comment]
Thank you again for coming and reviewing this PR, @rxin , @kiszk , @mridulm , @omalley .
So far, we discussed the followings.

1. `Why are we adding this to core? Why not just the hive module?` (@rxin)
   - `sql/core` module gives more benefit than `sql/hive`.
   - Apache ORC library (`no-hive` version) is a general and resonably small library designed for non-hive apps.

2. `Can we add smaller amount of new code to use this, too?` (@kiszk)
   - The previous #17980 , #17924, and #17943 are the complete examples containing this PR.
   - This PR is focusing on dependency only.

3. `Why don't we then create a separate orc module? Just copy a few of the files over?` (@rxin)
   -  Apache ORC library is the same with most of other data sources(CSV, JDBC, JSON, PARQUET, TEXT) which live inside `sql/core`
   - It's better to use as a library instead of copying ORC files because Apache ORC shaded jar has many files. We had better depend on Apache ORC community's effort until an unavoidable reason for copying occurs.

4. `I do worry in the future whether ORC would bring in a lot more jars` (@rxin)
   - The ORC core library's dependency tree is aggressively kept as small as possible. I've gone through and excluded unnecessary jars from our dependencies. I also kick back pull requests that add unnecessary new dependencies. (@omalley)

I tried to contain and summarize all advices here, but please let me know if I missed some concerns here.

**Comment by omalley** [comment]
I would also comment that in the long term, Spark should move to using the vectorized reader in ORC's core. That would remove the dependence on ORC's mapreduce module, which provides row by row shims on top of the vectorized reader.

**Comment by dongjoon-hyun** [comment]
Sure. Thank you so much, @omalley !

**Comment by dongjoon-hyun** [comment]
@rxin . Could you make some decision for this PR? Do we need to put this into `sql/hive` still for some reasons?

**Comment by dongjoon-hyun** [comment]
Retest this please.

**Comment by sameeragarwal** [comment]
LGTM; unless @rxin still has some strong objections?

**Comment by dongjoon-hyun** [comment]
Thank you so much, @sameeragarwal .

**Comment by dongjoon-hyun** [comment]
Hi, @mridulm, @sameeragarwal , and @rxin .
Please let me know if there is something for me to do here.

**Comment by dongjoon-hyun** [comment]
Retest this please.

**Comment by dongjoon-hyun** [comment]
Hi, @sameeragarwal and @mridulm .
I cannot see any clear reason for the objection here. Also, there is a positive feedback from @ash211 in the dev@spark, too. This PR will bring an improvement definitely. Could you merge this PR for Apache Spark to move forward?

File: pom.xml [hunk 1]
```diff
@@ -1678,6 +1681,44 @@
...
+        <classifier>${orc.classifier}</classifier>
+        <scope>${orc.deps.scope}</scope>
+        <exclusions>
+          <exclusion>
+            <groupId>org.apache.hadoop</groupId>
+            <artifactId>hadoop-common</artifactId>
+          </exclusion>
+          <exclusion>
+            <groupId>org.apache.hive</groupId>
+            <artifactId>hive-storage-api</artifactId>
```
Comment (cloud-fan) [inline pom.xml]: so the orc core module still contains hive related stuff?

File: sql/core/pom.xml [hunk 2]
```diff
@@ -87,6 +87,16 @@
     </dependency>
 
     <dependency>
+      <groupId>org.apache.orc</groupId>
+      <artifactId>orc-core</artifactId>
+      <classifier>${orc.classifier}</classifier>
```
Comment (cloud-fan) [inline sql/core/pom.xml]: sorry a dumb question, what does `classifier` mean? I don't see it in the rest of this pom file

**Comment by cloud-fan** [comment]
LGTM besides some minor questions, @rxin any more comments on this?

File: pom.xml (hunk 1 above)
Comment (cloud-fan) [inline pom.xml]: and to confirm, this exclusion is safe only if we don't use hive storage api of orc in sql/core, right?

File: sql/core/pom.xml (hunk 2 above)
Comment (rxin) [inline sql/core/pom.xml]: what exactly is the storage api? confused about this too ...

File: pom.xml (hunk 1 above)
Comment (dongjoon-hyun) [inline pom.xml]: Thank you so much for review, @cloud-fan .

- The original `orc-core-1.4.0.jar` has `hive-storage-api` dependency. ([Maven Repo](https://mvnrepository.com/artifact/org.apache.orc/orc-core/1.4.))
- `orc-core-1.4.0-nohive.jar` is a shaded jar file including `hive-storage-api` under `org.apache.orc` namespace.

`orc-core-1.4.0-nohive.jar` is designed for users and apps who don't want to depend on (or consider) `hive`. `nohive` is a classifier for this purpose.

This PR uses `orc-core-1.4.0-nohive` only. To avoid Maven confusion, this exclution makes it sure by removing the `hive-storage-api` dependency explicitly from `orc-core` artifact.

File: sql/core/pom.xml (hunk 2 above)
Comment (dongjoon-hyun) [inline sql/core/pom.xml]: In Maven, `classifier` allows to distinguish artifacts that were built from the same POM but differ in their content. Here, `nohive` is used and it refers `orc-core-1.4.0-nohive.jar` instead of `orc-core-1.4.0.jar`.

Comment (dongjoon-hyun) [inline sql/core/pom.xml]: Thank you, @rxin ! 
Currently, the `hive-storage-api` jar has something like `SearchArgument` and `PredicateLeaf`. Apache ORC is trying to become independent module like Parquet.

Comment (cloud-fan) [inline sql/core/pom.xml]: ok good to learn the `classifier` stuff... Does it work in SBT too?

Comment (dongjoon-hyun) [inline sql/core/pom.xml]: Yes. sbt understands `classifier` in pom file. Also, we can use classifier in sbt build file, too.

File: pom.xml [hunk 3]
```diff
@@ -1678,6 +1681,44 @@
...
+          </exclusion>
+        </exclusions>
+      </dependency>
+      <dependency>
+        <groupId>org.apache.orc</groupId>
+        <artifactId>orc-mapreduce</artifactId>
+        <version>${orc.version}</version>
+        <classifier>${orc.classifier}</classifier>
```
Comment (viirya) [inline pom.xml]: The `classifier` for `orc-mapreduce` is the same as `orc-core`?

Comment (dongjoon-hyun) [inline pom.xml]: Thank you for review, @viirya .
Yes, it's the same. There is `orc-mapreduce-1.4.0.jar` and `orc-mapreduce-1.4.0-nohive.jar`.

Comment (viirya) [inline pom.xml]: Thanks. I think they are come from https://issues.apache.org/jira/browse/ORC-174.

Comment (dongjoon-hyun) [inline pom.xml]: Right. The wording is a little bit different, but technically those jars come from that JIRA patch.

**Comment by viirya** [comment]
LGTM

**Comment by dongjoon-hyun** [comment]
Thank you again, @viirya .

File: sql/core/pom.xml (hunk 2 above)
Comment (omalley) [inline sql/core/pom.xml]: @rxin Storage-API is a separately released artifact from the Hive project. Basically, Storage-API are the in-memory format for Hive's vectorization. You could draw the analogy that Storage-Api is for Hive what Arrow is for Drill. It allows formats to read and write directly in the format that is needed by the execution engine.

With the nohive classifier, ORC shades the storage-api jar into the ORC namespace so that it is compatible with any version of Hive.

Comment (dongjoon-hyun) [inline sql/core/pom.xml]: Thank you, @omalley !

**Comment by dongjoon-hyun** [comment]
Hi, @cloud-fan , @rxin , @sameeragarwal and @mridulm .
Could you merge this PR?

**Comment by rxin** [comment]
lgtm

**Comment by dongjoon-hyun** [comment]
Thank you so much, @rxin , @cloud-fan , @sameeragarwal , @mridulm , @viirya !

**Comment by gatorsmile** [comment]
Thanks! Merging to master.

**Comment by dongjoon-hyun** [comment]
Thank you, @gatorsmile !!!
